_includes/css/activation.scss.liquid

```diff
@@ -45,288 +45,281 @@
+    To avoid page-dependent SCSS, all nav links initially have that background image.
+    The following rule removes the image from the links to all parents, siblings,
+    and children of the current page.
+  {%- endcomment %}
+
+    {% if nav_page_level >= 2 -%}
+    {%- for i in (1..nav_page_parent_level) -%}
+    {{ activation_collection_prefix }} >
+    {%- for j in (2..i) %} li > ul >
+    {%- endfor %} li > a,
+    {% endfor -%}
+    {%- endif %}
+    {{ activation_collection_prefix }} >
+    {%- for i in (1..nav_page_parent_level) %} li > ul >
+    {%- endfor %} li:not(:nth-child({{ nav_levels[nav_page_level] }})) > a,
+    {{ activation_collection_prefix }} >
+    {%- for i in (1..nav_page_level) %} li > ul >
+    {%- endfor %} li a {
+      background-image: none;
+    }
+
+  {%- comment -%}
+    The following rule removes the image from the links to pages in other collections.
+  {%- endcomment %}
+
+    {{ activation_other_collection_prefix }} a,
+    .site-nav li.external a {
+      background-image: none;
+    }
+
+  {%- comment -%}
+    The following rule styles the link to the current page.
+  {%- endcomment %}
+
+    {{ activation_collection_prefix }} > li:nth-child({{ nav_levels[1] }})
+    {%- for i in (2..nav_page_level) %} > ul > li:nth-child({{ nav_levels[i] }})
+    {%- endfor %} > a {
+      font-weight: 600;
+      text-decoration: none;
+    }
+
+  {%- comment -%}
+    The following rules unfold all collections, and display the links to any children
+    of the current page.
+    
+    To avoid dependence on the SCSS variable nav-list-expander-right, the direction
+    of the rotation of the expander icon is fixed, and corresponds to the appearance
+    when nav-list-expander-right is true. This results in a minor visual difference
+    between the appearance of active expander icons when JS is enabled/disabled and
+    nav-list-expander-right is false, which seems unavoidable.
+  {%- endcomment %}
+
+    {%- if site.just_the_docs.collections %}
+    .site-nav > ul.nav-category-list > li > button svg,
+    {% endif -%}
+    {{ activation_collection_prefix }} > li:nth-child({{ nav_levels[1] }}) > button svg
+    {%- for i in (2..nav_page_level) %},
+    {{ activation_collection_prefix }} > li:nth-child({{ nav_levels[1] }}) > 
+    {%- for j in (2..i) %} ul > li:nth-child({{ nav_levels[j] }}) >
+    {%- endfor %} button svg
+    {%- endfor %} {
+      transform: rotate(-90deg);
+    }
+
+    {%- if site.just_the_docs.collections %}
+    .site-nav > ul.nav-category-list > li.nav-list-item > ul.nav-list,
+    {% endif -%}
+    {{ activation_collection_prefix }} > li.nav-list-item:nth-child({{ nav_levels[1] }}) > ul.nav-list
+    {%- for i in (2..nav_page_level) %},
+    {{ activation_collection_prefix }} > li.nav-list-item:nth-child({{ nav_levels[1] }}) >
+    {%- for j in (2..i) %} ul.nav-list > li.nav-list-item:nth-child({{ nav_levels[j] }}) >
+    {%- endfor %} ul.nav-list
+    {%- endfor %} {
```

Comment:
Selectors that look like this are linear in the size of `nav_page_level`, right? I am wondering at what point CSS selector performance will become an issue. My guess is that most of our existing client sites won't have issues here (since they've been artificially limited to a depth of 2-3), but I'm curious how large sites with many more levels of nesting will do. I don't think that should be a barrier to implementing this change, but may be a future roadblock.

(I'm also cognizant that backwards-compatibility makes it even more challenging to implement this change well)

Reply:
> I'm curious how large sites with many more levels of nesting will do.

I think that sites with large numbers of pages tend to have high average branching. For 1000 pages with average branching 10, the average depth would probably be about 3. Even for a very slim site with 1000 pages and binary branching, the average depth wouldn't be much more than 10. Do you know of any sites with more than 10 navigation levels?

I know nothing about CSS selector efficiency. Would 10 levels of selectors make a website significantly less responsive? Note that most of the generated selectors are for direct descendants (using `>`) which ought to be quick to test.

In any case, the generated page-dependent CSS rules are _disabled_ when JS is active, so the vast majority of visitors would never be affected by any inefficiency with the selectors.

> (I'm also cognizant that backwards-compatibility makes it even more challenging to implement this change well)

Removing the restriction to 3 levels actually simplified the navigation link generation. I think the only tricky bit with backwards-compatibility was to treat a missing `grand_parent` field as indicating that the page indeed has no grandparent page when disambiguating parent pages.

Reply:
Thanks for the quick response, and apologies for how delayed this has been. I hadn't thought through the "average branching" idea - I agree with you that it's unlikely we would have large performance issues here. No change necessary!